### PR TITLE
feat(manager/helmfile): Lock file support

### LINKED ISSUE #7719: Helmfile Manager should update Lockfile
**What would you like Renovate to be able to do?**

The new Helmfile Manager currently only updates the version in the `helmfile.yaml`, but doesnt update the version and a sha256 digest in the `helmfile.lock
` file thats next to `helmfile.yaml`

**Did you already have any implementation ideas?**
The issue could be solved by installing `Helmfile` inside the Dockerfile and add a postUpdateOption that executes `helmfile deps` and commites those changes.

**Are there any workarounds or alternative ideas you've tried to avoid needing this feature?**
executing `helmfile deps` yourself and commit those changes in the MR branch

### PR DESCRIPTION
**Pull request by nikolaik**
## Changes

This PR attempts to add support for [helmfile](https://github.com/helmfile/helmfile) lock files. This PR is based on the existing work with helm chart lock files.

There is a corresponding PR to add the helmfile tool to the `renovate/renovate` docker image https://github.com/renovatebot/docker-renovate-full/pull/1276

## Context

We use renovate to keep our helmfile chart versions updated, so this is to avoid the toil of manually running `helmfile deps`.

Resolves #7719

Did an initial test run of this here: https://github.com/nikolaik/renovate-helmfile.lock/pull/2

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/helmfile/artifacts.spec.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    const updatedDeps = [{ depName: 'dep1' }];
+    expect(
+      await helmfile.updateArtifacts({
+        packageFileName: 'helmfile.yaml',
+        updatedDeps,
+        newPackageFileContent: chartFile,
+        config,
+      })
+    ).toBeNull();
+    expect(execSnapshots).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.spec.ts:87]: don't use snapshots, check other managers for samples

File: lib/util/exec/containerbase.ts [hunk 2]
```diff
@@ -65,6 +65,11 @@ const allToolConfig: Record<string, ToolConfig> = {
     depName: 'helm/helm',
     versioning: semverVersioningId,
   },
+  helmfile: {
```
Comment (viceice) [inline lib/util/exec/containerbase.ts:68]: this needs a new installer inside `containerbase/buildpack`[^1] image.

[^1]: https://github.com/containerbase/base

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/util/exec/containerbase.ts (hunk 2 above)
Comment (nikolaik) [reply lib/util/exec/containerbase.ts:68]: Made a PR for this in https://github.com/containerbase/base/pull/674

File: lib/modules/manager/helmfile/artifacts.spec.ts (hunk 1 above)
Comment (nikolaik) [reply lib/modules/manager/helmfile/artifacts.spec.ts]: Made an attempt at this in 7584677

File: lib/modules/manager/helmfile/artifacts.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    (updatedDeps === undefined || updatedDeps.length < 1)
+  ) {
+    logger.debug('No updated helmfile deps - returning null');
+    return null;
+  }
+
+  const lockFileName = getSiblingFileName(packageFileName, 'helmfile.lock');
+  const existingLockFileContent = await readLocalFile(lockFileName, 'utf8');
+
+  if (!existingLockFileContent) {
```
Comment (secustor) [inline lib/modules/manager/helmfile/artifacts.ts:42]:
```suggestion
  if (is.falsy(existingLockFileContent)) {
```

**Review by secustor — CHANGES_REQUESTED** [review]

File: lib/modules/manager/helmfile/artifacts.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            path: lockFileName,
+            contents: newHelmLockContent,
+          },
+        });
+      } else {
+        logger.debug('helmfile.lock is unchanged');
+      }
+    }
+
+    return fileChanges.length > 0 ? fileChanges : null;
```
[the remark this answers is not in the document]
Comment (nikolaik) [reply lib/modules/manager/helmfile/artifacts.ts]: Sure is, this was copied uncritically from the helm manager logic. Thanks!

File: lib/modules/manager/helmfile/__fixtures__/helmfile.lock [hunk 5]
```diff
@@ -0,0 +1,7 @@
+version: 0.149.0
```
Comment (secustor) [inline lib/modules/manager/helmfile/__fixtures__/helmfile.lock:1]: Please also add test cases in which the lock file is corrupted e.g. not valid yaml  and where it is containing multiple dependencies.

**Review by secustor — CHANGES_REQUESTED** [review]

File: lib/modules/manager/helmfile/__fixtures__/helmfile.lock (hunk 5 above)
Comment (nikolaik) [reply lib/modules/manager/helmfile/__fixtures__/helmfile.lock]: Found a few things that slipped through when trying to add the test cases you suggested. Code coverage for `lib/modules/manager/helmfile/artifacts.ts` is better now, but I'm unsure if the tests are testing the behaviour you describe. Do you have any pointers on how to test for the things you suggest?

File: lib/modules/manager/helmfile/artifacts.spec.ts [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+    ).toMatchObject([
+      {
+        file: {
+          type: 'addition',
+          path: 'helmfile.lock',
+          contents: lockFileTwo,
+        },
+      },
+    ]);
+    expect(execSnapshots).toBeArrayOfSize(1);
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.spec.ts:119]: you don't need this, toMatchObject ensures the right array length

File: lib/modules/manager/helmfile/artifacts.spec.ts [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    });
+    const updatedDeps = [{ depName: 'dep1' }];
+    expect(
+      await helmfile.updateArtifacts({
+        packageFileName: 'helmfile.yaml',
+        updatedDeps,
+        newPackageFileContent: helmfileYaml,
+        config,
+      })
+    ).toMatchObject([
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.spec.ts:146]:
```suggestion
    ).toEqual([
```
should work here

File: lib/modules/manager/helmfile/artifacts.spec.ts [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+          '-v "/tmp/github/some/repo":"/tmp/github/some/repo" ' +
+          '-v "/tmp/renovate/cache":"/tmp/renovate/cache" ' +
+          '-e BUILDPACK_CACHE_DIR ' +
+          '-e CONTAINERBASE_CACHE_DIR ' +
+          '-w "/tmp/github/some/repo" ' +
+          'renovate/sidecar ' +
+          'bash -l -c "' +
+          'install-tool helmfile v0.129.0' +
+          ' &&' +
+          ' helmfile deps -f helmfile.yaml' +
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.spec.ts:171]: please move the space to the ampersands

File: lib/modules/manager/helmfile/artifacts.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+import { exec } from '../../../util/exec';
+import type { ExecOptions, ToolConstraint } from '../../../util/exec/types';
+import {
+  getSiblingFileName,
+  readLocalFile,
+  writeLocalFile,
+} from '../../../util/fs';
+import type { UpdateArtifact, UpdateArtifactsResult } from '../types';
+
+async function helmfileCommand(
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.ts:14]: can you inline this function, as it's a simple online

File: lib/modules/manager/helmfile/artifacts.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+      constraint: config.constraints?.helmfile,
+    };
+
+    const execOptions: ExecOptions = {
+      docker: {},
+      extraEnv: {},
+      toolConstraints: [helmfileToolConstraint],
+    };
+    await helmfileCommand(execOptions, packageFileName);
+    logger.debug('Returning updated Helmfile artifacts');
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.ts:60]: remove

File: lib/modules/manager/helmfile/artifacts.ts [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+  const existingLockFileContent = await readLocalFile(lockFileName, 'utf8');
+
+  if (is.falsy(existingLockFileContent)) {
+    logger.debug('No helmfile.lock found');
+    return null;
+  }
+
+  try {
+    await writeLocalFile(packageFileName, newPackageFileContent);
+    logger.debug('Updating Helmfile artifacts');
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.ts:48]: remove

File: lib/modules/manager/helmfile/artifacts.ts [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+  await exec(`helmfile deps -f ${quote(helmfileYamlPath)}`, execOptions);
+}
+
+export async function updateArtifacts({
+  packageFileName,
+  updatedDeps,
+  newPackageFileContent,
+  config,
+}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
+  logger.debug(`helmfile.updateArtifacts(${packageFileName})`);
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.ts:27]:
```suggestion
  logger.trace(`helmfile.updateArtifacts(${packageFileName})`);
```

File: lib/modules/manager/helmfile/__fixtures__/helmfile-two.lock [hunk 13]
```diff
@@ -0,0 +1,10 @@
+version: 0.150.0
```
Comment (viceice) [inline lib/modules/manager/helmfile/__fixtures__/helmfile-two.lock:1]: please inline those fixtures and use `codeBlock`  string template for format help.

File: lib/modules/manager/helmfile/artifacts.spec.ts [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+
+jest.mock('../../datasource');
+jest.mock('../../../util/exec/env');
+jest.mock('../../../util/http');
+jest.mock('../../../util/fs');
+jest.mock('../../../util/git');
+
+const datasource = mocked(_datasource);
+
+const adminConfig: RepoGlobalConfig = {
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.spec.ts:21]: you need to set containerbase ENV to support install mode inside tests

File: lib/modules/manager/helmfile/artifacts.spec.ts [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+    fs.getParentDir.mockReturnValue('');
+    const updatedDeps = [{ depName: 'dep1' }, { depName: 'dep2' }];
+    expect(
+      await helmfile.updateArtifacts({
+        packageFileName: 'helmfile.yaml',
+        updatedDeps,
+        newPackageFileContent: helmfileYaml,
+        config,
+      })
+    ).toMatchObject([
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.spec.ts:110]:
```suggestion
    ).toEqual([
```

File: lib/modules/manager/helmfile/artifacts.spec.ts [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+    });
+    const updatedDeps = [{ depName: 'dep1' }];
+    expect(
+      await helmfile.updateArtifacts({
+        packageFileName: 'helmfile.yaml',
+        updatedDeps,
+        newPackageFileContent: helmfileYaml,
+        config,
+      })
+    ).toMatchObject([
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.spec.ts:196]:
```suggestion
    ).toEqual([
```

File: lib/modules/manager/helmfile/artifacts.spec.ts [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+          contents: lockFileTwo,
+        },
+      },
+    ]);
+    expect(execSnapshots).toMatchObject([
+      { cmd: 'helmfile deps -f helmfile.yaml' },
+    ]);
+  });
+
+  it('returns updated helmfile.lock with docker', async () => {
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.spec.ts:124]: add install binary source test

File: lib/modules/manager/helmfile/artifacts.ts [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+  if (
+    !isLockFileMaintenance &&
+    (updatedDeps === undefined || updatedDeps.length < 1)
+  ) {
+    logger.debug('No updated helmfile deps - returning null');
+    return null;
+  }
+
+  const lockFileName = getSiblingFileName(packageFileName, 'helmfile.lock');
+  const existingLockFileContent = await readLocalFile(lockFileName, 'utf8');
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.ts:32]: i think we should use `git.getFile` here, so we can better detect all changes

**Comment by nikolaik** [comment]
@viceice Worked my way through your comments. Let me know if I did something weird or misunderstood anything

File: lib/modules/manager/helmfile/artifacts.spec.ts (hunk 15 above)
Comment (viceice) [reply lib/modules/manager/helmfile/artifacts.spec.ts:147]: ☝️ ?

File: lib/modules/manager/helmfile/artifacts.spec.ts (hunk 16 above)
Comment (viceice) [reply lib/modules/manager/helmfile/artifacts.spec.ts:248]: ☝️ ?

**Comment by nikolaik** [comment]
Updated PR again @viceice Let me know if anything else looks weird

File: lib/modules/manager/helmfile/artifacts.ts [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+  if (is.falsy(existingLockFileContent)) {
+    logger.debug('No helmfile.lock found');
+    return null;
+  }
+
+  try {
+    await writeLocalFile(packageFileName, newPackageFileContent);
+    const helmfileToolConstraint: ToolConstraint = {
+      toolName: 'helmfile',
+      constraint: config.constraints?.helmfile,
```
Comment (secustor) [inline lib/modules/manager/helmfile/artifacts.ts:44]: Please add to the docs that now `helmfile` supports constraints. https://github.com/renovatebot/renovate/blob/main/docs/usage/configuration-options.md#constraints

**Review by secustor — CHANGES_REQUESTED** [review]

File: lib/modules/manager/helmfile/artifacts.ts (hunk 19 above)
Comment (nikolaik) [reply lib/modules/manager/helmfile/artifacts.ts:44]: Like this 55aa935 ?

**Review by secustor — DISMISSED** [review]

File: lib/modules/manager/helmfile/artifacts.spec.ts [hunk 20]
```diff
@@ -0,0 +1,257 @@
...
+      ).toEqual([
+        {
+          file: {
+            type: 'addition',
+            path: 'helmfile.lock',
+            contents: lockFileTwo,
+          },
+        },
+      ]);
+      expect(execSnapshots).toBeArrayOfSize(expectedCommands.length);
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.spec.ts:223]:
```suggestion
```
not needed, match below will do it

File: lib/modules/manager/helmfile/artifacts.spec.ts [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+    env.getChildProcessEnv.mockReturnValue(envMock.basic);
+    GlobalConfig.set(adminConfig);
+    docker.resetPrefetchedImages();
+    hostRules.clear();
+  });
+
+  afterEach(() => {
+    GlobalConfig.reset();
+  });
+
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.spec.ts:84]:
```suggestion
```
not needed, set above will overwrite all settings

File: lib/modules/manager/helmfile/artifacts.spec.ts [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+  repository: https://oauth2-proxy.github.io/manifests
+  version: 6.8.0
+digest: sha256:8ceea14d17c0f3c108a26ba341c63380e2426db66484d2b2876ab6e636e52af4
+generated: "2023-01-23T12:16:41.881988+01:00"
+`;
+
+describe('modules/manager/helmfile/artifacts', () => {
+  beforeEach(() => {
+    jest.resetAllMocks();
+    jest.resetModules();
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.spec.ts:73]: i think you can remove those

File: lib/modules/manager/helmfile/artifacts.spec.ts [hunk 23]
```diff
@@ -0,0 +1,257 @@
...
+        { cmd: 'install-tool helmfile v0.129.0' },
+        { cmd: 'helmfile deps -f helmfile.yaml' },
+      ],
+    },
+  ])(
+    'returns updated helmfile.lock with binarySource=$binarySource',
+    async ({ binarySource, expectedCommands }) => {
+      GlobalConfig.set({ ...adminConfig, binarySource });
+      fs.getSiblingFileName.mockReturnValueOnce('helmfile.lock');
+      git.getFile.mockResolvedValueOnce(lockFile as never);
```
Comment (viceice) [inline lib/modules/manager/helmfile/artifacts.spec.ts:196]: why the never?

**Review by secustor — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**